network/p2p/inspector/validation/control_message_validation_inspector.go

```diff
@@ -355,30 +389,49 @@ func (c *ControlMsgValidationInspector) inspectIHaveMessages(from peer.ID, ihave
 		Logger()
 	duplicateTopicTracker := make(duplicateStrTracker)
 	duplicateMessageIDTracker := make(duplicateStrTracker)
+	errs := make(p2p.InvCtrlMsgErrs, 0)
 	totalMessageIds := 0
 	for _, ihave := range ihaves {
 		messageIds := ihave.GetMessageIDs()
 		topic := ihave.GetTopicID()
 		if duplicateTopicTracker.isDuplicate(topic) {
-			return NewDuplicateTopicErr(topic, p2pmsg.CtrlMsgIHave), p2p.CtrlMsgNonClusterTopicType
+			errs = append(errs, p2p.NewInvCtrlMsgErr(NewDuplicateTopicErr(topic, p2pmsg.CtrlMsgGraft), p2p.ModerateErrSeverity, p2p.CtrlMsgNonClusterTopicType))
+			continue
 		}
 		duplicateTopicTracker.set(topic)
-		err, ctrlMsgType := c.validateTopic(from, channels.Topic(topic), activeClusterIDS)
+		err, topicType := c.validateTopic(from, channels.Topic(topic), activeClusterIDS)
 		if err != nil {
-			return err, ctrlMsgType
-		}
+			// short circuit validation when unstaked peer detected
+			if IsErrUnstakedPeer(err) {
+				lg.Debug().Msg("received control message from unstaked peer")
+				return nil
```

Comment:
⚠️ it seems this part needs reconsideration; if we keep _continuing_ the for loop, we never report the errors for unstated peers (on cluster prefix channels). Is it the desirable behavior?

Reply:
https://github.com/onflow/flow-go/pull/4978/commits/8a28bf97bad6e97fb921b27887f94b315c26d5f4